iceoryx2/src/port/publisher.rs

```diff
@@ -238,57 +234,110 @@ pub(crate) enum RemovePubSubPortFromAllConnectionsError {
 }
 
 #[derive(Debug)]
-pub(crate) struct DataSegment<Service: service::Service> {
+struct SegmentState {
     sample_reference_counter: Vec<IoxAtomicU64>,
-    memory: Service::SharedMemory,
-    payload_size: usize,
-    payload_type_layout: Layout,
+    payload_size: IoxAtomicUsize,
+}
+
+impl SegmentState {
+    fn new(number_of_samples: usize) -> Self {
+        let mut sample_reference_counter = Vec::with_capacity(number_of_samples);
+        for _ in 0..number_of_samples {
+            sample_reference_counter.push(IoxAtomicU64::new(0));
+        }
+
+        Self {
+            sample_reference_counter,
+            payload_size: IoxAtomicUsize::new(0),
+        }
+    }
+
+    fn set_payload_size(&self, value: usize) {
+        self.payload_size.store(value, Ordering::Relaxed);
+    }
+
+    fn payload_size(&self) -> usize {
+        self.payload_size.load(Ordering::Relaxed)
+    }
+
+    fn sample_index(&self, distance_to_chunk: usize) -> usize {
+        distance_to_chunk / self.payload_size()
```

Comment:
Would here a debug assert for `distance_to_chunk % self.payload_size() == 0` make sense?